src/lib/socket.c

```diff
@@ -2392,16 +2400,64 @@ static BIO_METHOD bio_methods = {
 	NULL,
 };
 
+static BIO_METHOD* get_bio_methods() {
+	return &bio_methods;
+} /* get_bio_methods() */
+#else
+CRYPTO_RWLOCK *bio_methods_lock = NULL;
+static CRYPTO_ONCE bio_methods_init = CRYPTO_ONCE_STATIC_INIT;
+static void bio_methods_lock_init(void) {
+	bio_methods_lock = CRYPTO_THREAD_lock_new();
+}
+static BIO_METHOD* get_bio_methods() {
+	static BIO_METHOD *bio_methods = NULL;
+	int type;
+
+	if (bio_methods == NULL)
+		return bio_methods;
+
+	if (!(CRYPTO_THREAD_run_once(&bio_methods_init, bio_methods_lock_init) ? bio_methods_lock != NULL : 0)) {
```

Comment:
It's kind of superfluous to use CRYPTO_THREAD_run_once to initialize a mutex, especially a mutex scoped for the current operation. Just move all the code into the routine being run once, and have it write to a global variable. That would also make it easier to specialize this stuff.